### PR TITLE
Docset Part 1: Support docset in config #396

### PR DESCRIPTION
**Pull request by superyyrrzz**
@chenkennt @vwxyzh @qinezh @hellosnow @ansyral 

sample docfx.json:
```json
{
  "build": {
    "docsets": {
      "v2.0": {
        "content": [
          {
            "files": "**/*.yml",
            "src": "v2.0",
            "dest": "api"
          }
        ],
        "globalMetadata": {
          "apiVersion": "v2.0"
        },
        "dest": "v2.0"
      },
      "v2.1": {
        "content": [
          {
            "files": "**/*.yml",
            "src": "v2.1",
            "dest": "api"
          }
        ],
        "globalMetadata": {
          "apiVersion": "v2.1"
        },
        "dest": "v2.1"
      }
    },
    "content": [
      {
        "files": [
          "*.md"
        ]
      }
    ],
    "dest": "_site"
  }
}
```

### CODE REVIEW

File: src/docfx/Models/BuildJsonConfig.cs [hunk 1]
```diff
@@ -96,5 +58,9 @@ public class BuildJsonConfig
         /// </summary>
         [JsonProperty("noLangKeyword")]
         public bool NoLangKeyword { get; set; }
+
+        [JsonProperty("docsets")]
+        public List<BuildJsonConfigCommon> Docsets { get; set; }
```
Comment (ansyral) [inline src/docfx/Models/BuildJsonConfig.cs]: need update our website's docfx.json?

**Comment by ansyral** [comment]
:shipit:

File: src/docfx/Models/BuildJsonConfig.cs (hunk 1 above)
Comment (superyyrrzz) [inline src/docfx/Models/BuildJsonConfig.cs]: As I only add a property "docset" property, it's compatible with the old docfx.json.

Comment (ansyral) [inline src/docfx/Models/BuildJsonConfig.cs]: but you moved some properties to `BuildJsonConfigCommon`, shouldn't the old docfx.json be updated to contain a `docsets` property with a list that has only one element?

**Comment by vwxyzh** [comment]
also update sample :)

**Comment by vwxyzh** [comment]
👍

File: Documentation/tutorial/docfx.exe_user_manual.md [hunk 2]
```diff
@@ -277,6 +278,26 @@ Please refer to [How to Create Custom Templates](howto_create_custom_template.md
 }
 ```
 
+#### 3.2.2 Properties for docset config
```
Comment (qinezh) [inline Documentation/tutorial/docfx.exe_user_manual.md]: Also add sample here?

**Comment by qinezh** [comment]
:shipit:

File: Documentation/tutorial/docfx.exe_user_manual.md [hunk 3]
```diff
@@ -277,6 +278,70 @@ Please refer to [How to Create Custom Templates](howto_create_custom_template.md
...
+
+*Docsets* are used to build multiple docsets with a single run of DocFX. The behavior is like running DocFX multiple times with different `docfx.json`. It's useful when building multiple versions of the same repository, which will lead to `uid` collusion if not put in different docsets.
+**NOTE** Cross Reference between different docsets is not allowed.
+
+Key                      | How to Merged with root config                                                                          | Description
```
Comment (hellosnow) [inline Documentation/tutorial/docfx.exe_user_manual.md]: merge

File: Documentation/tutorial/docfx.exe_user_manual.md [hunk 4]
```diff
@@ -277,6 +278,70 @@ Please refer to [How to Create Custom Templates](howto_create_custom_template.md
 }
 ```
 
+#### 3.2.2 Properties for docset config
+
+*Docsets* are used to build multiple docsets with a single run of DocFX. The behavior is like running DocFX multiple times with different `docfx.json`. It's useful when building multiple versions of the same repository, which will lead to `uid` collusion if not put in different docsets.
+**NOTE** Cross Reference between different docsets is not allowed.
```
Comment (hellosnow) [inline Documentation/tutorial/docfx.exe_user_manual.md]: I tried and found it's appending to previous line, add one line before to start a new line?

**Comment by hellosnow** [comment]
:shipit:

**Comment by superyyrrzz** [comment]
@jskeet To support multiple versions, now we have 2 designs of `docfx.json`:

Design 1. Add a dictionary named `versions` to the top level of `docfx.json`. The value of this dictionary is like a full `docfx.json` for a version. Outside `docsets`, all things remain the same:
```json
{
  "build": {
    "versions": {
      "v2.0": {
        "content": [
          {
            "files": "**/*.yml",
            "src": "v2.0",
            "dest": "api"
          }
        ],
        "dest": "v2.0"
      },
      "v2.1": {
        "content": [
          {
            "files": "**/*.yml",
            "src": "v2.1",
            "dest": "api"
          }
        ],
        "dest": "v2.1"
      }
    },
    "content": [
      {
        "files": [
          "*.md"
        ]
      }
    ],
    "dest": "_site"
  }
}
```
Design 2. A conciser way to solve uid conflict is to directly add `version` attribute in `content`. Files in each "version" have separated scope:
```json
{
  "build": {
    "content": [
      {
        "files": "*.yml",
        "src": "v2.0",
        "dest": "v2.0/api",
        "version": "v2.0"
      },
      {
        "files": "*.yml",
        "src": "v2.1",
        "dest": "v2.1/api",
        "version": "v2.1"
      },
      {
        "files": "conceptual/*.md",
        "src": "v2.0",
        "dest": "v2.0",
        "version": "v2.0"
      },
      {
        "files": "conceptual/*.md",
        "src": "v2.1",
        "dest": "v2.1",
        "version": "v2.1"
      }
    ],
    "dest": "_site"
  }
} 
```
From user's perspective, which one do you prefer?

**Comment by jskeet** [comment]
Option 2 looks a bit cleaner to me - but what's the impact of the two options beyond `docfx.json`? Two things I'm concerned about:

- How I refer to specific versions from top-level documentation - think of a home page with "Read the in-development documentation *here*, or the latest stable release *here*."
- How I refer to other items "within" a specific version, e.g. from a conceptual doc to a specific method.

In particular, for the latter, it would be a pain if that meant the doc code containing the version number - as moving from (say) v2.0 to v2.1 should start off by just copying the documentation source.

**Comment by superyyrrzz** [comment]
Thank you for your feedback:) We also tend to the later one. As both options aim to **separate** contents to avoid uid conflict, the impact of them are the same:

- For the first, due to the separated scope, one version can't refer to another. That's to say, DocFX can't resolve `[here](v2.0/index.md)` in top-level version, as it refers to a different version v2.0. A workaround is to use the correct relative path in output like`[here](v2.0/index.html)`. Reference cross versions still remains a problem that we will solve later.
- For the second, to refer to a specific method, you can use the method's uid directly. No extra version info need to be added as long as they have the same version in `docfx.json`.

**Comment by jskeet** [comment]
Great - that sounds fabulous to me. Will look forward to seeing how it progresses. Thanks very much for working on this :)

**Pull request closed without merging**